internal/route/always_route_filtering_strategy.go

```diff
@@ -0,0 +1,133 @@
+package route
+
+import (
+	"reflect"
+	"strings"
+
+	"github.com/satsuma-data/node-gateway/internal/config"
+	"github.com/satsuma-data/node-gateway/internal/metadata"
+	"github.com/satsuma-data/node-gateway/internal/types"
+	"go.uber.org/zap"
+)
+
+type AlwaysRouteFilteringStrategy struct {
+	BackingStrategy  RoutingStrategy
+	Logger           *zap.Logger
+	NodeFilters      []NodeFilter
+	RemovableFilters []NodeFilterType
+}
+
+func (s *AlwaysRouteFilteringStrategy) RouteNextRequest(
+	upstreamsByPriority types.PriorityToUpstreamsMap,
+	requestMetadata metadata.RequestMetadata,
+) (string, error) {
+	// Create a copy of removable filters to avoid modifying the original slice.
+	removableFilters := make([]NodeFilterType, len(s.RemovableFilters))
+	copy(removableFilters, s.RemovableFilters)
+
+	filters := s.NodeFilters // WARNING: Slice assignment, not a copy!
+	for len(filters) > 0 {
+		// Get all healthy upstreams for the current set of filters.
+		upstreams := s.FilterUpstreams(upstreamsByPriority, requestMetadata, filters)
+
+		// If there is at least one healthy upstream, route using the backing strategy.
+		if len(upstreams) > 0 {
+			return s.BackingStrategy.RouteNextRequest(upstreams, requestMetadata)
+		}
+
+		// There are no more filters to remove and no healthy upstreams found, so give up.
+		if len(removableFilters) == 0 {
+			break
+		}
+
+		// Remove the next filter and try again.
+		idx := len(removableFilters) - 1
+		nextFilterToRemove := removableFilters[idx]
+		removableFilters = removableFilters[:idx]
+		filters = removeFilters(filters, nextFilterToRemove)
+	}
+
+	// If all removable filters are exhausted and no healthy upstreams are found,
+	// pass in the original list of upstreams to the backing strategy.
+	//
+	// TODO(polsar): Eventually, we want all filters to be removable. Once that is the case,
+	// we should not be able to get here.
+	return s.BackingStrategy.RouteNextRequest(upstreamsByPriority, requestMetadata)
+}
+
+// FilterUpstreams filters upstreams based on the provided NodeFilter.
+// WARNING: If a given priority does not have any healthy upstreams,
+// it will not be included in the returned map.
+func (s *AlwaysRouteFilteringStrategy) FilterUpstreams(
```

Comment:
Doesn't this share a lot of code with [here](https://github.com/satsuma-xyz/node-gateway/blob/0e375d8e15f8c693e54be90c284f3ce29999e97f/internal/route/filtering_strategy.go#L24)? Is there a way to consolidate?

One idea is to to keep the logic of this current function, maybe put it in `filtering_strategy.go` as a static fn. Then it's trivial to call it from here. Then in the [existing](https://github.com/satsuma-xyz/node-gateway/blob/0e375d8e15f8c693e54be90c284f3ce29999e97f/internal/route/filtering_strategy.go#L24) `filter` function in `FilteringRoutingStrategy` , call this fn with `[]{s.NodeFilter}` as the 3rd argument.